patches/net/minecraft/world/inventory/AnvilMenu.java.patch

```diff
@@ -1,45 +1,48 @@
++        this.__createResultImpl();
++        ItemStack leftInput = this.inputSlots.getItem(0);
++        ItemStack rightInput = this.inputSlots.getItem(1);
++        net.neoforged.neoforge.common.CommonHooks.onAnvilUpdate(this, leftInput, rightInput, resultSlots, itemName, this.player);
++    }
 +
-         if (this.repairItemCountCost > 0) {
-             ItemStack itemstack = this.inputSlots.getItem(1);
-             if (!itemstack.isEmpty() && itemstack.getCount() > this.repairItemCountCost) {
-@@ -95,7 +_,7 @@
-         this.inputSlots.setItem(0, ItemStack.EMPTY);
-         this.access.execute((p_150479_, p_150480_) -> {
-             BlockState blockstate = p_150479_.getBlockState(p_150480_);
--            if (!p_150474_.hasInfiniteMaterials() && blockstate.is(BlockTags.ANVIL) && p_150474_.getRandom().nextFloat() < 0.12F) {
-+            if (!p_150474_.getAbilities().instabuild && blockstate.is(BlockTags.ANVIL) && p_150474_.getRandom().nextFloat() < breakChance) {
-                 BlockState blockstate1 = AnvilBlock.damage(blockstate);
-                 if (blockstate1 == null) {
-                     p_150479_.removeBlock(p_150480_, false);
-@@ -118,14 +_,20 @@
-         int i = 0;
-         long j = 0L;
-         int k = 0;
-+        if (!itemstack.isEmpty()) {  // don't fire the event if the left input slot is empty, because anvil shouldn't have recipes with an empty left slot
-+            if (!net.neoforged.neoforge.common.CommonHooks.onAnvilChange(this, itemstack, this.inputSlots.getItem(1), resultSlots, itemName, j, this.player)) {
-+                return;  // event is canceled or overrides the output item
-+            }
-+        }
-         if (!itemstack.isEmpty() && EnchantmentHelper.canStoreEnchantments(itemstack)) {
-             ItemStack itemstack1 = itemstack.copy();
-             ItemStack itemstack2 = this.inputSlots.getItem(1);
-             ItemEnchantments.Mutable itemenchantments$mutable = new ItemEnchantments.Mutable(EnchantmentHelper.getEnchantmentsForCrafting(itemstack1));
-             j += (long)itemstack.getOrDefault(DataComponents.REPAIR_COST, 0).intValue() + itemstack2.getOrDefault(DataComponents.REPAIR_COST, 0).intValue();
-             this.repairItemCountCost = 0;
-+            boolean flag = false;
-             if (!itemstack2.isEmpty()) {
--                boolean flag = itemstack2.has(DataComponents.STORED_ENCHANTMENTS);
-+                flag = itemstack2.has(DataComponents.STORED_ENCHANTMENTS);
-                 if (itemstack1.isDamageableItem() && itemstack.isValidRepairItem(itemstack2)) {
-                     int l2 = Math.min(itemstack1.getDamageValue(), itemstack1.getMaxDamage() / 4);
-                     if (l2 <= 0) {
++    // Neo: We patch the real createResult() into __createResultImpl() so we can wrap it, and provide the vanilla result in the AnvilUpdateEvent.
++    public void __createResultImpl() {
```

Comment:
I would remove the `__`. And maybe make the method protected if it's "impl"? Could also be `createResultInternal` if we want to bikeshed.

Your comment is also a bit weird. I'm referring to "we patch xxx into xxx." This is a pretty standard pattern. I'd rather write something like `Split actual result computation into a different method to be able to postprocess it in spite of the multiple returns`.

Reply:
Protected seems reasonable. Could add a `neo$` prefix to it to make it super obvious instead of `__`.

The comment is a bit weird, I was less focused on writing it and more focused on getting the whole thing spinning :p

Reply:
Just give it an idiomatic name. We don't prefix patched-in methods and I don't think we need to. We use a similar pattern in other places I think.

Reply:
Imagine `ItemTags.neo$create(RL)`... :P

Reply:
Updated